obkvrpc/compressor.go

```diff
@@ -0,0 +1,155 @@
+/*-
+ * #%L
+ * OBKV Table Client Framework
+ * %%
+ * Copyright (C) 2023 OceanBase
+ * %%
+ * OBKV Table Client Framework is licensed under Mulan PSL v2.
+ * You can use this software according to the terms and conditions of the Mulan PSL v2.
+ * You may obtain a copy of Mulan PSL v2 at:
+ *          http://license.coscl.org.cn/MulanPSL2
+ * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
+ * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
+ * See the Mulan PSL v2 for more details.
+ * #L%
+ */
+
+package obkvrpc
+
+import (
+	"bytes"
+	"compress/zlib"
+	"fmt"
+	"github.com/golang/snappy"
+	"github.com/klauspost/compress/zstd"
+	"github.com/oceanbase/obkv-table-client-go/protocol"
+	"github.com/pierrec/lz4"
+	"github.com/pkg/errors"
+	"io"
+)
+
+type DeCompressor interface {
+	Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error)
+}
+
+type NoneDecompressor struct{}
+type Lz4Decompressor struct{}
+type SnappyDecompressor struct{}
+type ZlibDecompressor struct{}
+type ZStdDecompressor struct{}
+
+func (d *NoneDecompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	return src, nil
+}
+
+func (d *Lz4Decompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	srcBytes := src.Bytes()
+	dstBytes := make([]byte, originLen)
+	n, err := lz4.UncompressBlock(srcBytes, dstBytes)
+	if err != nil {
+		return nil, err
+	}
+	if int32(n) != originLen {
+		return nil, errors.New(fmt.Sprintf("length of decompress content is not expected: actual-%d, expected-%d", n, originLen))
+	}
+	return bytes.NewBuffer(dstBytes), nil
+}
+
+func (d *SnappyDecompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	srcBytes := src.Bytes()
+	dstBytes, err := snappy.Decode(nil, srcBytes)
+	if err != nil {
+		return nil, err
+	}
+	if int32(len(dstBytes)) != originLen {
+		return nil, errors.New(fmt.Sprintf("length of decompress content is not expected: actual-%d, expected-%d", int32(len(dstBytes)), originLen))
```

Comment:
SnappyDecompressor: length of decompress content

Reply:
done